src/pool/pool_jemalloc.c

```diff
@@ -0,0 +1,345 @@
+
+    return arena_registry_g[arena_ind];
+}
+
+static void *arena_extent_alloc(extent_hooks_t *extent_hooks, void *new_addr,
+                                size_t size, size_t alignment, bool *zero,
+                                bool *commit, unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)new_addr;     // unused
+
+    jemalloc_memory_pool_t *pool = get_kind_by_arena(arena_ind);
+
+    void *ptr;
+    enum umf_result_t ret =
+        umfMemoryProviderAlloc(pool->hProvider, size, alignment, &ptr);
+    if (ret != UMF_RESULT_SUCCESS) {
+        return NULL;
+    }
+
+    *zero = true;
+    *commit = true;
+
+    return ptr;
+}
+
+static void arena_extent_destroy(extent_hooks_t *extent_hooks, void *addr,
+                                 size_t size, bool committed,
+                                 unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)committed;    // unused
+
+    jemalloc_memory_pool_t *pool = get_kind_by_arena(arena_ind);
+
+    enum umf_result_t ret = umfMemoryProviderFree(pool->hProvider, addr, size);
+
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "umfMemoryProviderFree failed \n");
+    }
+}
+
+static bool arena_extent_dalloc(extent_hooks_t *extent_hooks, void *addr,
+                                size_t size, bool committed,
+                                unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)committed;    // unused
+    (void)arena_ind;    // unused
+
+    return true;
+}
+
+static bool arena_extent_commit(extent_hooks_t *extent_hooks, void *addr,
+                                size_t size, size_t offset, size_t length,
+                                unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)offset;       // unused
+    (void)length;       // unused
+    (void)arena_ind;    // unused
+
+    return false;
+}
+
+static bool arena_extent_decommit(extent_hooks_t *extent_hooks, void *addr,
+                                  size_t size, size_t offset, size_t length,
+                                  unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)offset;       // unused
+    (void)length;       // unused
+    (void)arena_ind;    // unused
+
+    return true;
+}
+
+static bool arena_extent_purge_lazy(extent_hooks_t *extent_hooks, void *addr,
+                                    size_t size, size_t offset, size_t length,
+                                    unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)size;         // unused
+
+    jemalloc_memory_pool_t *pool = get_kind_by_arena(arena_ind);
+    enum umf_result_t ret = umfMemoryProviderPurgeLazy(
+        pool->hProvider, (char *)addr + offset, length);
+
+    return (ret == UMF_RESULT_SUCCESS);
```

Comment:
According to jemalloc [documentation](https://jemalloc.net/jemalloc.3.html): "If the function returns true, this indicates failure to purge." 

I think it is counterintuitive, but I didi the same mistake when created my initial prototype and after some time @igchor  pointed me to the documentation.

Reply:
It would be good to have a unit test that verifies that we implemented these callbacks correctly (according to the documentation).

Reply:
why did you mark it as resolved without any comments or code changes?

Reply:
I have changed the code a bit later. I do not want to push out each tiny change every time I address one comment, because each push generates many CI builds. When there is many comments it is much easier for me to mark it as resolved when I have fixed it on my branch (because it disappears then) and push out all changes when I fix all or most of comments. At the end I come back and unresolve all issues that seems not obvious.

Reply:
@vinser52 Done